### PR TITLE
Rewards

### PR DESCRIPTION
**Pull request by maldon**
Current workflow:
- user requests an asset snapshot at a specified block height:
`requestsnapshot asset_name block_height`
- user distributes reward
`distributereward asset_name snapshot_height distribution_asset_name gross_distribution_amount ( exception_addresses )`

### CODE REVIEW

File: src/assets/rewardrequestdb.cpp [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+)
+{
+    SetNull();
+
+    for (auto const & currReq : p_requests) {
+        requests.insert(currReq);
+    }
+}
+
+CRewardRequestDB::CRewardRequestDB(size_t nCacheSize, bool fMemory, bool fWipe) : CDBWrapper(GetDataDir() / "rewardrequest", nCacheSize, fMemory, fWipe) {
```
Comment (blondfrogs) [inline src/assets/rewardrequestdb.cpp]: I think we should wrap all rewards databases into the folder rewards
``` GetDataDir() / "rewards/rewardrequest"```

File: src/assets/rewardrequestdb.cpp [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    reqDbEntry.requests.insert(p_newReward);
+
+    for (auto const & reward : reqDbEntry.requests) {
+        LogPrintf("%s : Reward at height %d: rewardID='%s', wallet='%s', amt=%lld, srcAsset='%s', tgtAsset='%s', exceptions='%s'\n",
+            __func__,
+            reward.heightForPayout, reward.rewardID.c_str(),
+            reward.walletName.c_str(), static_cast<long long>(reward.totalPayoutAmt),
+            reward.payoutSrc.c_str(), reward.tgtAssetName.c_str(),
+            reward.exceptionAddresses.c_str());
+    }
```
Comment (blondfrogs) [inline src/assets/rewardrequestdb.cpp]: Need to make sure this doesn't make it passed testing. 
Could be good to put a ```//TODO - Remove this log loop before releasing```

File: src/assets/rewardrequestdb.cpp [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+            p_rewardID.c_str());
+
+        return false;
+    }
+    //  Retrieve height-based entry which contains list of rewards to pay at that height
+    CRewardRequestDBEntry reqDbEntry;
+
+    //  If this fails, bail, since we can't be sure we won't blow away valid entries
+    if (!Read(std::make_pair(SCHEDULEDREWARD_FLAG, rewardToRemove.heightForPayout), reqDbEntry)) {
+        LogPrintf("%s : Failed to read scheduled rewards at specified height!\n", __func__);
```
Comment (blondfrogs) [inline src/assets/rewardrequestdb.cpp]: This is another issue we run into when using sets that contain multiple items for the height as the value in the database

File: src/assets/rewardrequestdb.cpp [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+{
+    LogPrintf("%s : Looking for scheduled rewards at height %d!\n",
+        __func__, p_maxBlockHeight);
+
+    std::unique_ptr<CDBIterator> pcursor(NewIterator());
+
+    pcursor->SeekToFirst();
+
+    // Find out if any pending rewards exist at or below the specified height
+    while (pcursor->Valid()) {
```
Comment (blondfrogs) [inline src/assets/rewardrequestdb.cpp]: or below? Doesn't look like this covers that

File: src/assets/rewardrequestdb.cpp [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    }
+
+    for (auto const & reward : p_rewardRequests) {
+        LogPrintf("%s : Found payable reward: rewardID='%s', wallet='%s', height=%d, amt=%lld, srcAsset='%s', tgtAsset='%s', exceptions='%s'\n",
+            __func__,
+            reward.rewardID.c_str(),
+            reward.walletName.c_str(), reward.heightForPayout, static_cast<long long>(reward.totalPayoutAmt),
+            reward.payoutSrc.c_str(), reward.tgtAssetName.c_str(),
+            reward.exceptionAddresses.c_str());
+    }
```
Comment (blondfrogs) [inline src/assets/rewardrequestdb.cpp]: //TODO Make sure this doesn't get passed testing
We should create a LogPrint enum for rewards and use that to post to the debug log

File: src/rpc/rewards.cpp [hunk 6]
```diff
@@ -0,0 +1,883 @@
...
+    CAmount amount;
+    if (!ParseFixedPoint(p_value.getValStr(), 8, &amount))
+        throw JSONRPCError(RPC_TYPE_ERROR, strprintf("Invalid amount (3): %s", p_value.getValStr()));
+
+    if (p_isRVN && !MoneyRange(amount))
+        throw JSONRPCError(RPC_TYPE_ERROR, strprintf("Amount out of range: %s", amount));
+
+    return amount;
+}
+
```
Comment (blondfrogs) [inline src/rpc/rewards.cpp]: I think this function already exists. However, it doesn't take into the account of the p_isRVN bool. Maybe we should it, so we aren't duplicating code.

File: src/validation.cpp [hunk 7]
```diff
@@ -4996,10 +5033,25 @@ bool LoadBlockIndex(const CChainParams& chainparams)
...
+            fAssetIndex = true;
+            pblocktree->WriteFlag("assetindex", fAssetIndex);
+            LogPrintf("%s: asset index %s\n", __func__, fAssetIndex ? "enabled" : "disabled");
+        }
+        else {
+            // Use the provided setting for -assetindex in the new database
+            fAssetIndex = gArgs.GetBoolArg("-assetindex", DEFAULT_ASSETINDEX);
+            pblocktree->WriteFlag("assetindex", fAssetIndex);
+            LogPrintf("%s: asset index %s\n", __func__, fAssetIndex ? "enabled" : "disabled");
+        }
```
Comment (blondfrogs) [inline src/validation.cpp]: Just want to double check. 
So if rewards is enabled, it enables assetIndex.

File: src/init.cpp [hunk 8]
```diff
@@ -1599,18 +1624,24 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
                 if (!mapBlockIndex.empty() && mapBlockIndex.count(chainparams.GetConsensus().hashGenesisBlock) == 0)
                     return InitError(_("Incorrect or no genesis block found. Wrong datadir for network?"));
 
-                // Check for changed -txindex state
-                if (fTxIndex != gArgs.GetBoolArg("-txindex", DEFAULT_TXINDEX)) {
-                    strLoadError = _("You need to rebuild the database using -reindex to change -txindex");
+                // See if the state of the rewards system is being changed
+                if (fRewardsEnabled != gArgs.GetBoolArg("-rewards", DEFAULT_REWARDS_ENABLED)) {
+                    strLoadError = _("You need to rebuild the database using -reindex to change -rewards");
```
Comment (blondfrogs) [inline src/init.cpp]: This isn't exactly correct, if the user is just trying to turn off rewards right? 
I am thinking about it like so. If a wallet has assetindex enabled. They can use the rewards functionality. If they don't have it enabled, they can't use it. 
I do think there is a reason to have the -rewards, unless I am missing something. 
Let me know.

File: src/assets/assetdb.cpp [hunk 9]
```diff
@@ -391,4 +391,49 @@ bool CAssetsDB::AssetAddressDir(std::vector<std::pair<std::string, CAmount> >& v
...
+            bool excluded = false;
+            for (auto const & excludedAddr : p_excludedAddresses) {
+                if (key.second.second == excludedAddr) {
+                    excluded = true;
+                    break;
+                }
+            }
+
+            // If this is not an excluded address, include its amount in the total
+            if (!excluded) {
```
[the remark this answers is not in the document]
Comment (blondfrogs) [reply src/assets/assetdb.cpp]: Lets makes to not include burn addresses in the total.

File: src/assets/payoutdb.cpp [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+
+    //  There are still pending payments, so write them back to the DB
+    if (pendingPaymentsPresent) {
+        LogPrintf("UpdatePayoutEntry: Updating payout entry for reward '%s'\n", p_payoutEntry.rewardID.c_str());
+        return Write(std::make_pair(PAYOUT_FLAG, p_payoutEntry.rewardID), p_payoutEntry);
+    }
+
+    //  Otherwise, everything has been completed, so erase the DB entry
+    LogPrintf("UpdatePayoutEntry: Removing payout entry for reward '%s'\n", p_payoutEntry.rewardID.c_str());
+    return Erase(std::make_pair(PAYOUT_FLAG, p_payoutEntry.rewardID), true);
```
[the remark this answers is not in the document]
Comment (blondfrogs) [reply src/assets/payoutdb.cpp]: LogPrint(DEBUG::REWARDS, "UpdatePayoutEntry: Updating payout entry for reward '%s'\n", p_payoutEntry.rewardID.c_str());

File: src/assets/payoutdb.h [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        const std::string & p_rewardID,
+        CPayoutDBEntry & p_payoutEntry);
+
+    //  Remove the payout entry for the specified reward
+    bool RemovePayoutEntry(const std::string & p_rewardID);
+
+    //  Save the specified entry back to the database, including any modifications
+    bool UpdatePayoutEntry(const CPayoutDBEntry & p_payoutEntry);
+
+    bool Flush();
```
Comment (blondfrogs) [inline src/assets/payoutdb.h]: Lets just remove this.

File: src/assets/rewardrequestdb.h [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+    // Find out if any reward payments are scheduled at the specified height
+    bool AreRewardsScheduledForHeight(const int & p_blockHeight);
+
+    // Retrieve all reward records *for a specific asset* (if specified) at the provided block height
+    bool LoadPayableRewardsForAsset(
+        const std::string & p_assetName, const int & p_blockHeight,
+        std::set<CRewardRequest> & p_rewardRequests);
+
+    bool Flush();
```
Comment (blondfrogs) [inline src/assets/rewardrequestdb.h]: remove

File: src/rpc/rewards.cpp [hunk 13]
```diff
@@ -0,0 +1,883 @@
...
+                expectedCount++;
+
+                // Parse Raven address
+                CTxDestination dest = DecodeDestination(payment.address);
+                if (!IsValidDestination(dest)) {
+                    LogPrintf("Destination address '%s' is invalid.\n", payment.address.c_str());
+                    payment.completed = true;
+                    continue;
+                }
+
```
Comment (blondfrogs) [inline src/rpc/rewards.cpp]: Make sure all address are valid when creating the snapshot.

**Comment by maldon** [comment]
Still WIP implementing code review changes.

File: src/assets/assetdb.cpp [hunk 14]
```diff
@@ -391,4 +391,49 @@ bool CAssetsDB::AssetAddressDir(std::vector<std::pair<std::string, CAmount> >& v
...
+    p_totalAmount = 0;
+
+    while (pcursor->Valid()) {
+        boost::this_thread::interruption_point();
+
+        std::pair<char, std::pair<std::string, std::string> > key;
+        if (pcursor->GetKey(key) && key.first == ASSET_ADDRESS_QUANTITY_FLAG && key.second.first == p_assetName) {
+            //  Ensure that this is not one of the excluded addresses
+            bool excluded = false;
+            for (auto const & excludedAddr : p_excludedAddresses) {
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/assetdb.cpp]: This function was no longer needed so I've removed it.

File: src/assets/assetdb.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/assetdb.cpp]: Function was removed

File: src/assets/assetsnapshotdb.cpp [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+        return false;
+    }
+
+    //  Retrieve all of the addresses/amounts in batches
+    const int MAX_RETRIEVAL_COUNT = 100;
+    bool errorsOccurred = false;
+
+    for (int retrievalOffset = 0; retrievalOffset < totalEntryCount; retrievalOffset += MAX_RETRIEVAL_COUNT) {
+        //  Retrieve the specified segment of addresses
+        tempOwnersAndAmounts.clear();
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/assetsnapshotdb.cpp]: Done

File: src/assets/assetsnapshotdb.cpp [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+    }
+    if (ownersAndAmounts.size() == 0) {
+        LogPrintf("AddAssetOwnershipSnapshot: No owners exist for asset '%s'.\n", p_assetName.c_str());
+        return false;
+    }
+
+    //  Add them to the existing list
+    std::set<CAssetSnapshotDBEntry> snapshotEntries;
+
+    if (!Read(std::make_pair(SNAPSHOTCHECK_FLAG, p_height), snapshotEntries)) {
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/assetsnapshotdb.cpp]: The key is now a combined string using block height and asset name, and only one snapshot is stored per key.

File: src/assets/assetsnapshotdb.cpp [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+    std::set<CAssetSnapshotDBEntry> snapshotEntries;
+
+    //  Load up the snapshot entries at this height
+    if (!Read(std::make_pair(SNAPSHOTCHECK_FLAG, p_height), snapshotEntries)) {
+        return false;
+    }
+
+    //  Find out if this specific asset ownership entry for this height exists
+    for (auto const & entry : snapshotEntries) {
+        if (p_assetName == entry.assetName) {
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/assetsnapshotdb.cpp]: The key is now a combined string using block height and asset name, and only one snapshot is stored per key.

File: src/assets/assetsnapshotdb.cpp [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+        pcursor->SeekToFirst();
+
+        // Find out if any pending rewards exist at or below the specified height
+        while (pcursor->Valid()) {
+            boost::this_thread::interruption_point();
+            std::pair<char, int> key;
+
+            //  Retrieve entries at the current node and remove this asset, if present
+            if (pcursor->GetKey(key) && key.first == SNAPSHOTCHECK_FLAG) {
+                std::set<CAssetSnapshotDBEntry> snapshotEntries;
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/assetsnapshotdb.cpp]: Resolved by key change and storage of a single entry per key.

File: src/assets/assetsnapshotdb.cpp [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+
+bool CAssetSnapshotDB::RemoveAssetFromEntries(
+    const std::string & p_assetName, int p_height,
+    std::set<CAssetSnapshotDBEntry> & p_snapshotEntries)
+{
+    //  Find out if this specific asset ownership entry for this height exists
+    std::set<CAssetSnapshotDBEntry>::iterator entryIT;
+
+    for (entryIT = p_snapshotEntries.begin(); entryIT != p_snapshotEntries.end(); ) {
+        if (entryIT->assetName == p_assetName) {
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/assetsnapshotdb.cpp]: Resolved by key change and storage of a single entry per key.

File: src/assets/assetsnapshotdb.cpp [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+            entryIT = p_snapshotEntries.erase(entryIT);
+        }
+        else {
+            ++entryIT;
+        }
+    }
+
+    //  If it is non-empty, write it back out
+    if (p_snapshotEntries.size() > 0) {
+        return Write(std::make_pair(SNAPSHOTCHECK_FLAG, p_height), p_snapshotEntries);
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/assetsnapshotdb.cpp]: Resolved by key change and storage of a single entry per key.

File: src/assets/payoutdb.cpp [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+    //      fail the distribution, since it is not possible to reward all ownerships equally.
+    if (srcIsIndivisible
+        && (
+            p_rewardReq.totalPayoutAmt < p_assetSnapshot.totalAmountOwned   //  Fail, not enough
+            || p_rewardReq.totalPayoutAmt % p_assetSnapshot.totalAmountOwned != 0   //  Fail, can't distribute evenly
+        )
+    ) {
+        LogPrintf("GeneratePayouts: Source asset '%s' is indivisible, and not enough reward value was provided for all ownership of '%s'\n",
+            p_rewardReq.payoutSrc.c_str(), p_rewardReq.tgtAssetName.c_str());
+        return false;
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/payoutdb.cpp]: Correct, user can retry with different amount or payout source.

File: src/assets/payoutdb.cpp [hunk 22]
```diff
@@ -0,0 +1,273 @@
...
+    for (entryIT = nonExceptionPayments.begin(); entryIT != nonExceptionPayments.end(); ) {
+        bool isExceptionAddress = false;
+
+        for (auto const & exceptionAddr : exceptionAddressList) {
+            if (entryIT->address == exceptionAddr) {
+                isExceptionAddress = true;
+                break;
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/payoutdb.cpp]: This file was removed, but the code in question has been updated to use a set of exception addresses.

File: src/assets/payoutdb.cpp [hunk 23]
```diff
@@ -0,0 +1,273 @@
...
+    //  Update data in the provided payout entry
+    p_payoutEntry.rewardID = p_rewardReq.rewardID;
+    p_payoutEntry.assetName = p_rewardReq.tgtAssetName;
+    p_payoutEntry.srcAssetName = p_rewardReq.payoutSrc;
+
+    for (auto & payment : nonExceptionPayments) {
+        //  Calculate the reward amount
+        double floatingCurrAssetAmt = payment.payoutAmt;
+        double floatingRewardAmt = floatingPaymentAmt * floatingCurrAssetAmt / floatingTotalAssetAmt;
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/payoutdb.cpp]: Done

File: src/assets/payoutdb.cpp [hunk 24]
```diff
@@ -0,0 +1,273 @@
...
+    if (nonExceptionPayments.size() == 0) {
+        LogPrintf("GeneratePayouts: Ownership of '%s' includes only exception/burn addresses.\n",
+            p_rewardReq.tgtAssetName.c_str());
+        return false;
+    }
+
+    //  Loop through asset owners
+    double floatingPaymentAmt = p_rewardReq.totalPayoutAmt;
+    double floatingTotalAssetAmt = p_assetSnapshot.totalAmountOwned;
+
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/payoutdb.cpp]: Done

File: src/assets/payoutdb.cpp [hunk 25]
```diff
@@ -0,0 +1,273 @@
...
+    //  After saving our payout entry, remove the reward request
+    if (!pRewardRequestDb->RemoveReward(p_rewardReq.rewardID)) {
+        //  Not critical if this fails
+        LogPrintf("GeneratePayouts: Failed to remove reward request for reward '%s'\n", p_rewardReq.rewardID.c_str());
+    }
+
+    //  Find out if any reward requests are still pending for this asset at the current height
+    if (!pRewardRequestDb->AreRewardsScheduledForHeight(p_rewardReq.heightForPayout)) {
+        //  Delete the asset snapshot
+        if (!pAssetSnapshotDb->RemoveOwnershipSnapshot(p_rewardReq.tgtAssetName, p_rewardReq.heightForPayout)) {
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/payoutdb.cpp]: All automatic removal of records has been removed. RPC commands exist to manually remove records.

File: src/assets/payoutdb.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/payoutdb.cpp]: New debug option, rewards, has been added and used for all Rewards-specific log messages.

File: src/assets/rewardrequestdb.cpp (hunk 1 above)
Comment (maldon) [reply src/assets/rewardrequestdb.cpp]: Done

File: src/assets/rewardrequestdb.cpp (hunk 2 above)
Comment (maldon) [reply src/assets/rewardrequestdb.cpp]: This block has been removed.

File: src/assets/rewardrequestdb.cpp (hunk 3 above)
Comment (maldon) [reply src/assets/rewardrequestdb.cpp]: Resolved by changing key to a string concatenation of block height and asset name.

File: src/assets/rewardrequestdb.cpp (hunk 4 above)
Comment (maldon) [reply src/assets/rewardrequestdb.cpp]: Fixed comment by removing the "or below" portion, which was no longer applicable.

File: src/assets/rewardrequestdb.cpp (hunk 5 above)
Comment (maldon) [reply src/assets/rewardrequestdb.cpp]: This block has been removed/changed and now only lists block height and asset name.

File: src/rpc/rewards.cpp (hunk 6 above)
Comment (maldon) [reply src/rpc/rewards.cpp]: Done

File: src/validation.cpp (hunk 7 above)
Comment (maldon) [reply src/validation.cpp]: The rewards enablement flag has been removed.

File: src/init.cpp (hunk 8 above)
Comment (maldon) [reply src/init.cpp]: The rewards enablement flag has been removed.

File: src/assets/payoutdb.h (hunk 11 above)
Comment (maldon) [reply src/assets/payoutdb.h]: Done

File: src/assets/rewardrequestdb.h (hunk 12 above)
Comment (maldon) [reply src/assets/rewardrequestdb.h]: Done

File: src/rpc/rewards.cpp (hunk 13 above)
Comment (maldon) [reply src/rpc/rewards.cpp]: Ownership addresses are now validated during snapshot creation.

File: src/assets/assetsnapshotdb.h [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+        assetName = "";
+        totalAmountOwned = 0;
+        ownersAndAmounts.clear();
+    }
+
+    bool operator<(const CAssetSnapshotDBEntry &rhs) const
+    {
+        std::string myHeightAndAsset = std::to_string(height) + assetName;
+        std::string theirHeightAndAsset = std::to_string(rhs.height) + rhs.assetName;
+        return myHeightAndAsset < theirHeightAndAsset;
```
[the remark this answers is not in the document]
Comment (maldon) [reply src/assets/assetsnapshotdb.h]: The concatenated block height + asset name string is still the key, but it is generated in the constructor, so hopefully the impact is minimal. This combined value just works really well as the key.

**Comment by maldon** [comment]
Ready for review again. I marked all of the conversations as resolved, but I'm now not sure if I was supposed to : )

Something to note, I didn't dynamically enable the ConnectTip functionality that checks for pending snapshot requests, as I couldn't tell if this would persist when the daemon was restarted. So this functionality is currently on by default now that the rewards enablement flag has been removed.
I could use some direction on how to implement this properly.

**Comment by maldon** [comment]
Ready for review again. I marked all of the conversations as resolved, but I'm now not sure if I was supposed to : )

Something to note, I didn't dynamically enable the ConnectTip functionality that checks for pending snapshot requests, as I couldn't tell if this would persist when the daemon was restarted. So this functionality is currently on by default now that the rewards enablement flag has been removed.
I could use some direction on how to implement this properly.

**Comment by blondfrogs** [comment]
@maldon Thanks. I will take a look. Lets talk more about your ConnectTip question.

**Comment by maldon** [comment]
@blondfrogs was there more that needed to be done here? More functional testing maybe?

**Comment by blondfrogs** [comment]
@maldon you could do something like this 
git rebase -i HEAD~#NUMBEROFCOMMITS

Then you would assign almost all of these commits to be squashed, and comment out the commit message. 

I can show you if you want to try it out

**Comment by maldon** [comment]
I'll give that a shot. Also, my python tests are failing now, I assume because something changed in the main code:
`2019-09-04 14:47:25.862000 TestFramework (ERROR): JSONRPC error
Traceback (most recent call last):
  File "/home/maldon/src/ravencoin/test/functional/test_framework/test_framework.py", line 127, in main
    self.run_test()
  File "test/functional/feature_rewards.py", line 397, in run_test
    self.basic_test_rvn()
  File "test/functional/feature_rewards.py", line 67, in basic_test_rvn
    n0.transfer(asset_name="STOCK1", qty=200, to_address=shareholderAddr0, change_address=ownerAddr0)
  File "/home/maldon/src/ravencoin/test/functional/test_framework/coverage.py", line 48, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/maldon/src/ravencoin/test/functional/test_framework/authproxy.py", line 142, in __call__
    raise JSONRPCException(response['error'])
test_framework.authproxy.JSONRPCException: JSON value is not a string as expected (-1)`

**Comment by maldon** [comment]
@blondfrogs  - okay, code is rebased and squashed. I'm still seeing the same failure in my python test, but I don't know what is causing it.

**Comment by maldon** [comment]
@blondfrogs python test is fixed, ready for review

**Pull request closed without merging**